src/app/app.component.ts

```diff
@@ -599,69 +638,81 @@ export class AppComponent implements OnInit, AfterViewInit, CanvasTableSelectLis
     const snackBarRef = this.snackBar.open('Toggling read status...');
     const messageIds = this.canvastable.rows.selectedMessageIds();
 
-
-    const args = {
-      flag: {
-        name: 'seen_flag',
-        value: status ? 1 : 0,
-      },
-      ids: messageIds
-    };
-
-    this.rmm.email.update(args).subscribe(() => {
-        this.searchService.updateIndexWithNewChanges();
-        if (!this.searchService.localSearchActivated) {
-          messageIds.forEach( (id) => {
-               this.rmmapi.messageFlagChangeSubject.next(
-                   new MessageFlagChange(id, status, null)
-               );
-          } );
-        }
+    this.messageActionsHandler.updateMessages(
+      messageIds,
+      // local index / messagelist
+      (msgIds: number[]) => {
+        msgIds.forEach( (id) => {
+          // Updates both index + messagelist
+          this.rmmapi.messageFlagChangeSubject.next(
+            new MessageFlagChange(id, status, null)
+          );
+        });
         this.clearSelection();
-        this.showSelectMarkOpMenu = false;
-        snackBarRef.dismiss();
-    });
+      },
+      // remote api
+      (msgIds: number[]) => {
+        const req = this.rmm.email.update({
+          flag: {
+            name: 'seen_flag',
+            value: status ? 1 : 0,
+          },
+          ids: msgIds
+        });
+        req.subscribe(() => snackBarRef.dismiss());
```

Comment:
This may be a good fit for a `.pipe(tap(() => snackBarRef.dismiss()))` instead of a `subscribe()`, which would ensure that it doesn't get executed twice (even if it weren't `shared()`).